AMAROC-768 add heigh cloud flag and allow NSC

regDaniel opened the pull request:
**DRAFT: Add high clouds Flag**

1. Add a high clouds flag to data in case we discard more hits above MSA than `MAX_HITS_OKTA0`
2. Set NSC instead of NCD in case the above flag is True when we create the METAR message.

**Checklists**:
- [x] New code includes dedicated tests.
- [x] New code has been linted.
- [x] New code follows the project's style.
- [x] New code is compatible with the 3-Clause BSD license.
- [x] CHANGELOG has been updated.
- [x] AUTHORS has been updated.
- [x] Copyright years in module docstrings have been updated.

fpavogt commented on src/ampycloud/data.py:
```diff
@@ -1036,6 +1074,6 @@ def metar_msg(self, which: str = 'layers') -> str:
 
         # Here, deal with the situations when all clouds are above the MSA
         if len(msg) == 0:
```
I just realized that we also need to take into account the significance of the layers located between the MSA and MSA+Delta MSA. So the update should look more like something along the following lines:

```
layers_in_msa_buffer = len(sligrolay['significant']*(sligrolay['alt_base'] >= msa_val)) > 0. # Before l. 1072

...

return 'NSC' if layers_in_msa_buffer else self._ncd_or_nsc() . # To replace l. 1077
```

The last line may not be the best, because the code is hardcoded ?

regDaniel replied:
I think it's fine to hardcode this as this code will very likely not change. I implemented it along your sugestion.

fpavogt commented on src/ampycloud/data.py:
```diff
@@ -93,16 +94,31 @@ def _cleanup_pdf(self, data: pd.DataFrame) -> pd.DataFrame:
         # Begin with a thorough inspection of the dataset
         data = utils.check_data_consistency(data, req_cols=self.DATA_COLS)
 
+        # By default we set NSC to false, NOTE: attrs is still experimental in pandas!
+        try:
+            data.attrs['high_clouds_detected'] = False
+        except AttributeError:
+            # NOTE: Untestable as we cannot monkeypatch attrs as it would result in side-effects for __repr__
+            # which break other tests
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore") # pandas assumes we are trying to assign a column, which is not the case
+                data.attrs = {'high_clouds_detected': False}
+
```
Interesting approach ! I like it, in that it keeps this very close to the data. Without thinking too much about it, I guess I would have implemented it as a class property (e.g. like self.msa). But if you're ok with this approach, then it is certainly ok with me.

regDaniel replied:
I'm okay with it and prefer to have the flag close to the data (reason: I think the CeiloChunk class is quite big and versatile already, we might want to split it up at some point and it will come in handy if there are not more attributes).

fpavogt commented on test/ampycloud/test_data.py:
```diff
@@ -228,6 +253,36 @@ def test_ceilochunk_nocld():
+    chunk.find_layers()
+
+    # Assert the final METAR code is correct
+    assert chunk.metar_msg() == 'NSC'
```
Could be good to also add a test when there is a cloud layers within the MSA buffer, just to make sure we issue 'NSC' in this case as well.

regDaniel replied:
I added one here by parameterizing the mock layers, one within thr buffer and one above.

fpavogt reviewed (changes requested):
Looks good @regDaniel. I think a small tweak is needed to handle the case when there is a significant layer (or more) in the region between the MSA and the MSA + Delta MSA

regDaniel commented:
> Looks good @regDaniel. I think a small tweak is needed to handle the case when there is a significant layer (or more) in the region between the MSA and the MSA + Delta MSA

absolutely. good catch, thanks! I'll take your suggestions and will think about how to best implement it.

regDaniel commented:
Hi @fpavogt , thanks a lot for your review! I implemented the search for layers in buffer along your suggestions and added a test. Could you double-check if it's fine?

fpavogt reviewed (approved):
Looks good to me.

regDaniel commented on .github/workflows/CI_pylinter.yml:
```diff
@@ -32,9 +32,10 @@ jobs:
+      - name: Install dependencies
         run: |
           python -m pip install --upgrade pip
+          python -m pip install setuptools
```
Thanks! Will take a note to remove it again during the rebasing process.

loforest commented on src/ampycloud/data.py:
```diff
@@ -93,16 +94,33 @@ def _cleanup_pdf(self, data: pd.DataFrame) -> pd.DataFrame:
             # non-detection below the MSA. Also change the hit type to 0 accordingly !
-            data.loc[data[(data.alt > hit_alt_lim) & (data.type <= 1)].index, 'type'] = 0
-            data.loc[data[(data.alt > hit_alt_lim) & (data.type <= 1)].index, 'alt'] = np.nan
+            above_msa_t1_or_less = data[(data.alt > hit_alt_lim) & (data.type <= 1)].index
```
I am confused about this part: data.type <= 1 includes -1 (VV hit), 0 (no hit) and 1 (first cloud base). I do not understand why we separate data this way (type [-1, 0, 1] at line 115 vs type [2, 3] at line 119)

It was already like this in the previous version, but I am wondering if we are doing things correctly. Using -1 for VV might be more error prone when using "lower than" conditions. Use 4 for VV hit instead? (tbd with @fpavogt)

fpavogt replied:
The motivation for using -1 for VV hits is that some ceilometers may possibly return more than 3 simultaneous base heights per time step (either now, or in the future). I find using -1 much safer from this perspective, because it is clearly "unique".

The separation at line 115 comes from the fact that we cannot simply "remove" the cloud base detections above the MSA from the dataset: we need to replace them with a non-detection (else, we would over-estimate the density of the other layers). To replace a cloud detection with a non-detection, we must set its type to 0 - but only once per time step. For cases with multiple base detections (with type 1, 2, 3), a single entry must be turned to type 0, and all the others must be cropped from the dataset (else, we would have created duplicated hits of type 0).

regDaniel replied:
Nothing to add - I agree with the reasoning as to why set -1 for VV hits. I'll add a comment on why we are changing types.

loforest commented on src/ampycloud/data.py:
```diff
@@ -361,12 +380,11 @@ def _setup_sligrolay_pdf(self, which: str = 'slices') -> tuple[pd.DataFrame, npt
 
         # We want to raise early if 'which' is unknown.
-        if not which in ['slices', 'groups', 'layers']:
+        if which not in ['slices', 'groups', 'layers']:
```
Was this a bug or does it have the same behaviour?

fpavogt replied:
Not a bug and the behavior was identical, but the new version is better Python syntax.

regDaniel replied:
yes, sorry, I first thought the pipeline failed because of too many pylint issues, so I went for some easy ones. But then I realized it was due to `setuptools` missing as a dependency with Python 3.12. Nevertheless, I'd keep these little style adjustment to comply better with PEP standards.

loforest commented on src/ampycloud/data.py:
```diff
@@ -992,6 +1010,28 @@ def layers(self) -> pd.DataFrame:
 
+    @property
+    def high_clouds_detected(self) -> bool:
+        """ Returns whether high clouds were detected in the data.
```
```suggestion
        """ Returns whether clouds above the MSA were detected in the data.
```

regDaniel replied:
```suggestion
        """ Returns whether clouds above (MSA + MSA_HIT_BUFFER) were detected in the data.
```

loforest commented on src/ampycloud/data.py:
```diff
@@ -992,6 +1010,28 @@ def layers(self) -> pd.DataFrame:
 
+    @property
+    def high_clouds_detected(self) -> bool:
+        """ Returns whether high clouds were detected in the data.
```
How should this be interpreted exactly? 
No cloud hits above the MSA? Above the MSA + buffer? (before running ampycloud)
No cloud layer base above the MSA? (after running ampycloud)
We should write the docstring to clarify.

regDaniel replied:
see suggestion above.

loforest commented on src/ampycloud/data.py:
```diff
@@ -992,6 +1010,28 @@ def layers(self) -> pd.DataFrame:
+    def _ncd_or_nsc(self) -> str:
+        """ Return the METAR code for No Cloud Detected / No Significant Cloud.
+        Decision based on the attribute set in data.attrs.
+
```
```suggestion

Note that regulation says that "if there are no clouds of operational significance and no restriction on vertical visibility and the abbreviation ‘CAVOK’ is not appropriate, the abbreviation ‘NSC’ should be used" (AMC1 MET.TR.205(e)(1)). Therefore, the NSC returned by ampycloud only covers the case of no clouds of operational significance (except CB and TCU). 

```

loforest replied:
I think we should also/only put this sentence (or similar) into the "Scope of ampycloud" section here: https://meteoswiss.github.io/ampycloud/index.html

fpavogt replied:
I would propose the following tweak to this proposition, to also include the limitations of the 'NCD'. I would agree to have this in the online docs, and I'll also add it to the scientific article.

```
Note that regulation says that "if there are no clouds of operational significance and no restriction on vertical visibility and the abbreviation ‘CAVOK’ is not appropriate, the abbreviation ‘NSC’ should be used" (AMC1 MET.TR.205(e)(1)). ampycloud cannot decide whether a 'CAVOK' is appropriate, and will therefore always return 'NSC' if no clouds of operational significance are found. If no clouds are detected at all by the ceilometers, ampycloud will return 'NCD'. Importantly, users should bear in mind that ampycloud cannot handle CB and TCU cases, such that any 'NCD'/'NSC' codes issued may need to be overwritten by the user in certain situations. 
```

regDaniel replied:
added it to `scope.rst`

loforest commented on src/ampycloud/data.py:
```diff
@@ -1036,6 +1076,9 @@ def metar_msg(self, which: str = 'layers') -> str:
         if len(msg) == 0:
-            return 'NCD'
+            sligrolay_in_buffer = sligrolay['significant'] * (sligrolay['alt_base'] >= msa_val)
+            if sligrolay_in_buffer.any():  # clouds within the MSA buffer zone
+                return 'NSC'
+            return self._ncd_or_nsc()
```
To check if I understood correctly: 

- line 1080 checks whether there are any cloud layer bases above the MSA (after running ampycloud)
- line 1082 checks whether there are any ceilometer hits above the MSA (before running ampycloud)

I am a bit confused and additional comments into the code would help.

regDaniel replied:
added a comment.

loforest reviewed (commented):
I reviewed the code and added some questions to ensure I understand the code (some docstrings might be extended accordingly) and also proposed a more explicit variable name for high_clouds_detected (to be discussed).

regDaniel commented on src/ampycloud/data.py:
```diff
@@ -93,16 +94,33 @@ def _cleanup_pdf(self, data: pd.DataFrame) -> pd.DataFrame:
 
+        # By default we set NSC to false, NOTE: attrs is still experimental in pandas!
+        try:
+            data.attrs['high_clouds_detected'] = False
```
Hmm, we can rename the flag, but `clouds_above_msa` would be misleading in my opinion because this flag is set for clouds above MSA + MSA buffer.

regDaniel replied:
named it `clouds_above_msa_buffer` - ok?

fpavogt replied:
Sounds good to me.

regDaniel commented on src/ampycloud/data.py:
```diff
@@ -93,16 +94,33 @@ def _cleanup_pdf(self, data: pd.DataFrame) -> pd.DataFrame:
         # Begin with a thorough inspection of the dataset
         data = utils.check_data_consistency(data, req_cols=self.DATA_COLS)
 
+        # By default we set NSC to false, NOTE: attrs is still experimental in pandas!
```
well yes. As @fpavogt mentions, we could just set a `CeiloChunk` attribute. But then I think CeiloChunk is already to big which is why I'd like to keep the flag close to the data. And there it's the only way. I tried defining a specific `AmpycloudDataFrame` class, but it would lead to having to convert everything you read in with the pandas library. It's a bit hard to explain, we can chat if you need clarification.

fpavogt replied:
Maybe the idea of setting up an `AmpycloudDataFrame` could be added to the todo list ? Doing so could be a good way to stabilize the API in terms of the data input by further shielding the user from the inner-workings. And allow for exploring the potential of polars to replace pandas, which could/should lead to even better performances (cf #96) ? Such a replacement would however be a rather big surgery. I would advise against doing it right away/in one go.

regDaniel replied:
Yes! I'd like to look into polars at some point. But I do agree that this would be a lot of work.

regDaniel commented on CHANGELOG:
```diff
@@ -3,6 +3,8 @@ All notable changes to ampycloud will be documented in this file.
 This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [v1.1.0]
+ - [regDaniel, 2024-04-09] Add flag for high clouds and allow for NSC in METAR message
```
```suggestion
 - [regDaniel, 2024-04-09] Add flag clouds above (MSA + MSA_HIT_BUFFER) and allow for NSC in METAR message
```
